### ISSUE TITLE
Change parser/render api

### ISSUE THREAD

**Issue by voischev** [posthtml/posthtml#198]
Expample:
```js
var postHTML = function(plugins) { this.plugins = plugins; };

postHTML.prototype.parser = function(html) { return 'OLD' + html; }; // CHANGE
postHTML.prototype.process = function(html, options) {
  var tree = this.parser(html);
  if (!this.plugins.length)
    return tree;
  
  this.plugins.forEach(function(plug) {
    tree = plug(tree);
  });
  
  return tree;
  
}

var ph = function(plugins) {
  return new postHTML(plugins);
}

var pDefault = ph([function(s) { return s + 'DEFAULT_PLUGINS'; }]);
var pCustom = ph([function(s) { return s + 'CUSTOM_PLUGINS'; }]);


pCustom.parser = function customParser(html) {return html + 'NEW'; };

console.log(pDefault.parser('sss'));
console.log(pDefault.process('sss'));

console.log(pCustom.parser('sss'));
console.log(pCustom.process('sss'));

console.log('\n\n\nchain', pDefault.process(pCustom.process('sss')));
```

**Comment by michael-ciniawsky** [comment]
```js
const sml = require('posthtml-sml')(options)

// ?
poshtml(plugins)
  .parser(sml)
  .process(index.sml, { from: '', to: '' })
  .render()
```

👍 

```js
PostHTML.prototype.parser
PostHTML.prototype.render 
```

👎 

> Delete options.parser/options.render

What is disliked setting them via options in `posthtml.process` could this not be an either?

```js
parser = this.parser || options.parser
render = this.render || options.render
```

**Comment by michael-ciniawsky** [comment]
```js
// posthtml().parser(html) => result.tree
PostHTML.prototype.parser = function (carser, html) {
  this.parser = carser ? carser : parser

  if (html && typeof html === 'string') {
    return this.parser(html)
  }

  return this
}

// posthtml().render(tree) => result.html
PostHTML.prototype.render = function (cender, tree) {
  this.render = cender ? cender : render

  if (tree && typeof tree === 'object') {
    return this.render(tree)
  }

  return this
}

...

options = options || {}

 if (options.parser && typeof options.parser === 'function') {
   this.parser = options.parser
 }
 if (options.render && typeof options.render === 'function') {
   this.render = options.render
 }

tree = typeof tree === 'object' ? tree : this.parser(tree)

tree.parser = parser // posthtml-parser
tree.render = render // posthtml-render

tree.plugins = this.plugins
tree.options = options

tree.messages = []
```

`this.tree` ?

```
posthtml().parser(sml, index.sml).process()
```

But I'm confused by this change and I don't understand the examples :)

```js
// s/Default/Custom Plugins ?
var pDefault = ph([function(s) { return s + 'DEFAULT_PLUGINS'; }]);
var pCustom = ph([function(s) { return s + 'CUSTOM_PLUGINS'; }]);

// New/Old ?
postHTML.prototype.parser = function(html) { return 'OLD' + html; };

pCustom.parser = function customParser(html) {return html + 'NEW'; };
```

**Comment by tunnckoCore** [comment]
@michael-ciniawsky :+1: 

@voischev inspiration can be @PostCore (which I will soon update to use @node-MiniBase)_ and @wooorm's Unified. I also already mention my view in some issue. And I believe it would be the best (not because it is my view, it is proven).

**Comment by michael-ciniawsky** [comment]
@voischev Can you elaborate on your parser/render API changes once more please ? :)

@tunnckoCore What is node-minibase/node-base, I looked at node-base a while ago but I don't get what it is meant for and how to utilise it :)

**Comment by voischev** [comment]
> more please

upd example

@michael-ciniawsky 
chaining parser/render for what? :-)
```js
poshtml(plugins)
  .parser(sml)
  .parser(xml)
  .parser(yml)
  .parser(zml) // ?))
  .render(jsx) // ...
  .render(jsx2) // ?
  .render(jsx3) // :-)
  .process(index.sml, { from: '', to: '' }) 
```

declare parser for instance (in my example) looks better

**Comment by michael-ciniawsky** [comment]
@voischev Yeah no... :), thats why I'm asking because I don't get how the proposed change should/will look like. I'm still in favour of `options.parser/options.render`

### PostHTML

```js
// What is OLD?
PostHTML.prototype.parser = function(html) { return 'OLD' + html; }; // CHANGE
```

### Usage 

If options.parser/options.render is removed, how does an Node API example look like?

```js
const posthtml = require('posthtml')

posthtml(plugins)
    // ?
    .then((result) => console.log(result.html))
```

**Comment by tunnckoCore** [comment]
I don't like these APIs for few reasons. And really like the job that @jonschlinkert and @doowb created for Base's API because its concept and idea of "smart plugins" is just amazing and awesome.

@michael-ciniawsky Base is absolutely small and minimal "framework" for building robust tools - there already huge ecosystem around it and few tools built with it - Verb, Templates, Generate, Assemble, Update. I very familiar with it, all the components and concepts behind it. I'm following Jon and know deeply almost all of its ~1000 packages.

And because I know how Base is built and on top of what and how the things work, I decided to decouple it more in a bit more small parts. So MiniBase is more minimal than Base and can be used to create the Base framework - need to push few more plugins to be fully compatible.

So what's the "smart plugins" concept. It is absolutely simple and easy.

- Best to know is that they are synchronous in first place.
- They are functions that are called immediately when instance is created.
- They are defined using `.use` method - think `.use((app, base) => {})`
- Where that `app` argument passed to them is the instance of the "Base app" - think for that as "tool" created using Base
- Second argument passed to that function is `base` which is the very first instance of Base (its useful for Base app that creates other Base apps - for example Assemble uses Assemble Core, it uses Generate, it uses Templates, it uses Base; or Verb uses Generate, which again uses Base) - nevermind, that is a bit advanced thing and you don't need it so much to make everything works out of the box
- Usually that function passed to `.use` (immediately called) is used for extending App's APIs - adding more methods and properties to the instance and etc
- Then if that function returns another function, that functions is collected to one `fns` array (plugins)
- These collected "actual plugins" are not called until you call ".run" method
- That `.run` method is used to pass whatever you want to that `plugin` function as arguments
- In order to have that `.run` method you should include, ironically, the plugin `base-plugins` :D which overrides the internal `.use` method and adds that `.run` method

The diffs between Base and MiniBase are that Base has more methods and internal deps - and for reason. But in context of PostHTML for example you won't need most of the things in Base - it has `.set`, `.get` methods and etc - that's why I created MiniBase - liked the idea and how inheritance is done. So the way to go here is MiniBase - not because it is mine, but because it's better fit in that case. I already started few tools around it (postcore, postjson, limonjs, docksjs, gruu, asiajs and couple more ideas) - same as Base appeared. I just need 1-2 plugins to do to be compat with Base and when I'm ready I'll PR there with variant that uses minibase + plugins which will pass all of the Base tests.

```js
var Base = require('base')
var plugins = require('base-plugins')
var myApp = new Base()

myApp.use(plugins())
myApp.use(function (app) {
  // app === myApp === this
  console.log(app.define)
  console.log(this.define)
  console.log(app.use)
  console.log(this.use)

  return function (file) {
    // file === this (maybe, not sure, but nevermind)
    console.log(file.path)
    console.log(file.contents)
  }
})

myApp.run({
  foo: 'bar',
  path: './a/b/c/ddd.txt',
  contents: 'foo bar welcome and hello'
})
```

It's simply amazing to know and use that ecosystem.

So what we can do in Post* tools is simple. We can do cool and flexible API using same approach. I think I already suggested how it can look like. Passing parser/render through `options` is what I did in PostCore too. And I don't see other way - it's the best.

Think for that

```js

// Array `plugins`
// Object `options`
//
// where `plugins` can also be an object with `.plugins` prop
posthtml(plugins, options)

// first style
// `.process` will run the plugins
// and it uses internally combo
// of `.parse` and `.render` methods
posthtml(plugins, options).process('str', options)

// second
posthtml(options)
  .use(smartPlugin)
  .use(smartPlugin)
  .use(smartPlugin)
  .process('str', options)

// third, just for someone
// that just want to parse
var astTree = posthtml(options).parse('str', options)

// fourth, just render given AST
// if AST is skipped it will be smart enough
// to check `this.ast` (or something) from previous
// `.parse` call; that's how `postcore` works
var strOutput = posthtml(options).render(astTree, options)

```

So yea, again, `.process` will use `.parse` and `.render` methods internally and the very special thing that will do is to "run" the plugins. Where that running will transform given string to a file objec. And here comes the cool part, we can use `VFile` to get `messages`, reports and sourcemaps for example for free.

That's how it will look

```js

posthtml(options)
  .use((app) => (file) => {
    // file will have all the great propeties
    // of the VFile - `.messages`, `.basename`
    // `.message` method, `.fail` method and etc
    console.log(file.message)
    console.log(file.fail)
    console.log(file.path) // not useful in PostHTML context but
    console.log(file.basename) // not useful in PostHTML context but
    console.log(file.messages)

    // will represent original string
    console.log(file.contents)

    // it will contain any additional data
    // such as AST, options; it can be in `file`
    // directly, but it's recommended to be in `.data`
    console.log(file.data)
  })
  .process('some awesome html', options)
```

Even more, parser/renderers can be plugins - so not only possible to pass through `options` object, but it can be passed directly as plugin - and that's how PostCore currently works.

And one note about `.process`. I completely don't see value to return a promise. Not only don seeing value and not only in PostHTML, but PostCSS too. I don't see sense, because parsing/compiling/rendering is completely synchronous thing. Tokenizer behind PostCSS too. All plugins, both PostHTML and PostCSS plugins, are totally synchronous, why is that returning a Promise. Absolute non-sense for me. It would be understandable if at least plugins can be async, but they aren't.

**Comment by michael-ciniawsky** [comment]
@tunnckoCore First thy for the detailed explanation of (mini) Base, it seems to be a opinionated convinience wrapper for prototypical Inheritance. But I'm uncertain if it is really needed for  posthtml.

```js
const tree = posthtml().parse('html', options)
const html = posthtml().render(tree, options)
```
👍 

Vfile/Vinyl is interesting for file metadata/messages/souremaps, I though about it a while ago, investigating into that is definitely not a bad idea . But the current initial messages support, which is simply an array with API methods to push formatted messages, also does the job( not to be meant as the best/final approach, by no means :) ). So the question is 'is it really necessary?'. Maybe for sourcemaps support? Do you know if Vfile/Vinyl has support for generating sourcemaps 'out-of-the-box'* (e.g 'vinyl-sourcemaps', 'vfile-sourcemaps' etc). So the sourcemaps implementation could be 'outsourced' to VfIle/Vinyl?

* Regardless of what 'processor'  does transform the contents, just input => output tracking with correct mappings for debugging ?

> Even more, will be able to use eslint reporters :) 

👍  The reporter style/format of posthtml-reporter should go into that direction, but we need messages first :). 

> Not only don't seeing value and not only in PostHTML, but PostCSS too; but I don't see any 
> sense, because parsing/compiling/rendering is completely synchronous thing. Tokenizer 
> behind PostCSS too. All plugins, both PostHTML and PostCSS plugins, are totally synchronous, > why is that returning a Promise. Absolute non-sense for me. It would be understandable if at 
> least plugins can be async, but they aren't

Can you elaborate on that futher please, I understand your concern to the point that without `fs/net` etc.  envolved there is no/not much benefit in async, since other transformations will be blocking/running in the main thread anyways. e.g eslint is 100% sync to my knowledge.

```js
export default function (options = {}) {
   return function posthtmlPlugin (tree) {
     // Sync
     if (tree.sync) {
       ...
       // 'bad'
       fs.readFileSync(...)
     }
     // Async
     ...
     const promises = []
     // 'good'
     promises.push(new Promise(resolve, reject) => { fs.readFile(...) })
      
     return promises.length > 0 ? Promise.all(promises).then(() => tree) : tree
   }
}
```

e.g Is the async part in the example not async for whatever reason I'm to dump for ;)?

**Comment by tunnckoCore** [comment]
> wrapper for prototypical Inheritance.

Not only. Plugins are powerful and done right. There's a way to detect if plugin is called already - "is registered" thing. Ensuring that each plugin will be called only once. Which is great benefit.

For example, there are some plugins that uses other plugins. So when you create a plugin your first job is to add `if (app.isRegistered('my-plugin-name')) return`, so when your plugin is included from another plugin it won't be called two times if it is already registered.

Such plugins are [base-task-alias](https://github.com/tunnckoCore/base-task-alias) and [base-task-render](https://github.com/tunnckoCore/base-task-render/blob/master/index.js#L71-L74) for example. Where in `base-task-alias` I need to be sure that some methods exists in case user that uses that plugin doesn't add them before adding my plugin. In other hand, when user include them before my plugin, my plugin won't call them again - it's just guaranteed that they will be called only once. So the case with `base-task-render` i included `base-task`, `base-fs`, `assemble-render-file` and `base-fs-conflicts`.

Base/minibase is just robust system with not big weight, giving good foundation to start up.


> So the question is 'is it really necessary?'. Maybe for sourcemaps support?

Yea. Reinventing the wheel is not necessary. You trying to do exactly this for me :) "Messages" will come for free when just convert incoming html string to vfile/vinyl files. And in that case, vfile is better.

> Can you elaborate on that futher please

I don't knwo what can say more. It's not a problem still to have both sync and async approaches. It would be just through options. For example `options.async: true` and `.process` will return a promise, otherwise it will return `result` object with `.code` and `.map` properties. Actually, would make more sense to be the file object, because plugins will work on it.

**Comment by michael-ciniawsky** [comment]
@tunnckoCore

> Not only. Plugins are powerful and done right.

yep of course, just 'summed up' my current understanding of minibase/base no offense 😛

> Yea. Reinventing the wheel is not necessary. You trying to do exactly this for me :) "Messages" will come for free when just convert incoming html string to vfile/vinyl files. And in that case, vfile is better.

I don't fully remember how in detail Vfile/Vinyl did it, but yep that should work with either of them.
Something like that. If sourcemaps are easily possible with Vfile/Vinyl, that approach should supersede the current primitive `{Array}`. Otherwise the extra dependency is unecessary.

```js
PostHTML.prototype.process = function (tree, options) {
   const file = new Vfile(tree, { messages: ... })
 
   ... parse

   tree.file = file 
}
```

make PoC (Gist/PR) for Messages with Vfile/Vinyl  :) 👍 

> I don't knwo what can say more. It's not a problem still to have both sync and async approaches. > It would be just through options. For example options.async: true and .process will return a 
> promise, otherwise it will return result object with .code and .map properties. Actually, would 
> make more sense to be the file object, because plugins will work on it.

Ok , I though there was something buggy which neglects asynchronicity somehow.

If `options.sync` or `options.async` by default. `options.async` is better because imho e.g `posthtml-include/posthtml-content` with fs interaction are likely to be in the setup.

**Comment by jonschlinkert** [comment]
Just wanted to say I think posthtml is great, and it would be awesome to see wider adoption of [base](https://github.com/node-base/base). But I certainly understand if you decide to go in a different direction.

fwiw, the main advantage of base is that it greatly simplifies creating and using composable plugins that are written mostly in vanilla javascript. The fact that base doesn't "do much" is really what's so appealing about it (IMHO). Base stays mostly agnostic to your conventions and coding preferences, it's main purpose is just to provide very lightweight scaffolding around what is otherwise entirely your own application.
